### PR TITLE
Generator op and auto minibatch scaling

### PR DESCRIPTION
**Pull request by ferrine**
This PR implements auto minibatching with appropriate likelihood scaling. Now generator can be passed to data.

### CODE REVIEW

**Comment by ferrine** [comment]
There is one but significant problem with generators. Any exception within calling `__next__` can will break the generator inside `theano.function`

File: pymc3/model.py [hunk 1]
```diff
@@ -458,7 +458,7 @@ def cont_vars(self):
         """All the continuous variables in the model"""
         return list(typefilter(self.vars, continuous_types))
 
-    def Var(self, name, dist, data=None):
+    def Var(self, name, dist, data=None, population=None):
```
Comment (twiecki) [inline pymc3/model.py]: I wonder if `population` is the best name. `population_size`, `total_size`?

File: pymc3/model.py [hunk 2]
```diff
@@ -717,7 +720,7 @@ class FreeRV(Factor, TensorVariable):
     """Unobserved random variable that a model is specified in terms of."""
 
     def __init__(self, type=None, owner=None, index=None, name=None,
-                 distribution=None, model=None):
+                 distribution=None, model=None, population=None):
```
Comment (twiecki) [inline pymc3/model.py]: Needs good doc-string.

File: pymc3/tests/test_theanof.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import unittest
+import numpy as np
+import theano
+from ..theanof import DataGenerator, GeneratorOp, generator
+
+
+def integers():
+    i = 0
+    while True:
+        yield np.float32(i)
```
Comment (twiecki) [inline pymc3/tests/test_theanof.py:11]: I don't think the casting should be necessary here?

File: pymc3/tests/test_theanof.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        for ndim in range(10):
+            res = list(itertools.islice(integers_ndim(ndim), 0, 2))
+            generator = DataGenerator(integers_ndim(ndim))
+            gop = GeneratorOp(generator)()
+            f = theano.function([], gop)
+            self.assertEqual(ndim, res[0].ndim)
+            np.testing.assert_equal(f(), res[0])
+            np.testing.assert_equal(f(), res[1])
+
+    def test_cloning_available(self):
```
Comment (twiecki) [inline pymc3/tests/test_theanof.py:44]: Neat!

**Comment by twiecki** [comment]
This is really neat!

What happens if the generator raises an exception?

**Comment by ferrine** [comment]
@twiecki then generator dies, I need to handle that case in some way or say user that it is dangerous. I think the latter is the only way I can solve the problem.

File: pymc3/model.py (hunk 1 above)
Comment (ferrine) [reply pymc3/model.py]: I like both `population` and `total_size`. `total size` is probably more intuitive

**Comment by ferrine** [comment]
I can check if generator is picklable as such generators don't die when exception is caught

**Comment by ferrine** [comment]
the second option is [here](http://www.fiber-space.de/generator_tools/doc/generator_tools.html) (one more dependency and not always works)

**Comment by twiecki** [comment]
"generator dies" -- raises another exception, crashes the kernel? how does it look like exactly?

**Comment by ferrine** [comment]
@twiecki after exception generator raises `StopIteration` when `__next__` is called, it becomes exhausted

**Comment by twiecki** [comment]
I see, and in general this should never happen, the generators should be non-exhaustive. Thus, when that happens we can raise an exception that says that the generator stopped for unknown reasons and that we can't tell whether it was an exception. That seems acceptable to me.

**Comment by ferrine** [comment]
`KeyboardInterrupt` can be annoing

**Comment by twiecki** [comment]
Well, that's triggered by the user so it should be fairly obvious what happened in that case.

**Comment by ferrine** [comment]
he will need to recompile the whole model

**Comment by twiecki** [comment]
I don't think we need to stress over this. The generator will most likely be fast so the probability that the KeyboardInterrupt would happen there is low.

**Comment by ferrine** [comment]
I'm stressed as risk is potentially high. Imagine DNN as a model. I would be disappointed

**Comment by twiecki** [comment]
OK, how about we do not raise an exception but just display a warning. Question is what to do next, I suppose the only way to exit gracefully would be to raise a `KeyboardInterrupt` that will be caught by the inference algorithm which would be super odd. Or we add another Exception case to the inference to exist gracefully.

**Comment by twiecki** [comment]
Is that a general constraint with generators? That exceptions do not percolate up?

**Comment by ferrine** [comment]
These exceptions cannot be caught. If they appear inside `__next__` it is fail in any case. `KeyboardInterrupt` is an exception that cannot be avoided

**Comment by twiecki** [comment]
Well, that's a limitation of the language then. I think we should move ahead with raising an Exception of the generator stops. If it becomes annoying later on we can think of other options. I really don't think it's all that critical.

**Comment by ferrine** [comment]
Good documentation is needed so

**Comment by twiecki** [comment]
Agreed, documentation will mention that.

Do you instead want to merge this into your VI branch so that we can test how this would work in practice?

**Comment by ferrine** [comment]
Yes, I'll add documentation and then merge

**Pull request closed without merging**